commit 6ccc971ee2c61a1ffb487e46bf6184f7df6aacfb
Author: Marc Zyngier <maz@kernel.org>
Date:   Fri Apr 19 11:29:32 2024 +0100

    KVM: arm64: nv: Add emulation for ERETAx instructions
    
    FEAT_NV has the interesting property of relying on ERET being
    trapped. An added complexity is that it also traps ERETAA and
    ERETAB, meaning that the Pointer Authentication aspect of these
    instruction must be emulated.
    
    Add an emulation of Pointer Authentication, limited to ERETAx
    (always using SP_EL2 as the modifier and ELR_EL2 as the pointer),
    using the Generic Authentication instructions.
    
    The emulation, however small, is placed in its own compilation
    unit so that it can be avoided if the configuration doesn't
    include it (or the toolchan in not up to the task).
    
    Reviewed-by: Joey Gouly <joey.gouly@arm.com>
    Reviewed-by: Oliver Upton <oliver.upton@linux.dev>
    Link: https://lore.kernel.org/r/20240419102935.1935571-13-maz@kernel.org
    Signed-off-by: Marc Zyngier <maz@kernel.org>

diff --git a/arch/arm64/include/asm/kvm_nested.h b/arch/arm64/include/asm/kvm_nested.h
index dbc4e3a67356..5e0ab0596246 100644
--- a/arch/arm64/include/asm/kvm_nested.h
+++ b/arch/arm64/include/asm/kvm_nested.h
@@ -64,4 +64,16 @@ extern bool forward_smc_trap(struct kvm_vcpu *vcpu);
 
 int kvm_init_nv_sysregs(struct kvm *kvm);
 
+#ifdef CONFIG_ARM64_PTR_AUTH
+bool kvm_auth_eretax(struct kvm_vcpu *vcpu, u64 *elr);
+#else
+static inline bool kvm_auth_eretax(struct kvm_vcpu *vcpu, u64 *elr)
+{
+	/* We really should never execute this... */
+	WARN_ON_ONCE(1);
+	*elr = 0xbad9acc0debadbad;
+	return false;
+}
+#endif
+
 #endif /* __ARM64_KVM_NESTED_H */